### PR TITLE
[SPARK-10248] [core] track exceptions in dagscheduler event loop in tests

### PR DESCRIPTION
**Pull request by squito**
`DAGSchedulerEventLoop` normally only logs errors (so it can continue to process more events, from other jobs).  However, this is not desirable in the tests -- the tests should be able to easily detect any exception, and also shouldn't silently succeed if there is an exception.

This was suggested by @mateiz on https://github.com/apache/spark/pull/7699.  It may have already turned up an issue in "zero split job".

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 1]
```diff
@@ -700,7 +700,7 @@ class DAGScheduler(
 
   private[scheduler] def cleanUpAfterSchedulerStop() {
     for (job <- activeJobs) {
-      val error = new SparkException("Job cancelled because SparkContext was shut down")
+      val error = new SparkException(s"Job ${job.jobId} cancelled because SparkContext was shut down")
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: not necessary, but was helpful for some debugging, figured it couldn't hurt.

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 2]
```diff
@@ -1150,7 +1150,7 @@ class DAGScheduler(
       case TaskResultLost =>
         // Do nothing here; the TaskScheduler handles these failures and resubmits the task.
 
-      case other =>
+      case ExecutorLostFailure(_) | TaskKilled | UnknownReason =>
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: this change is so I could trigger an exception (by passing in `null`), seems like a good change in any case.

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 3]
```diff
@@ -260,13 +267,18 @@ class DAGSchedulerSuite
...
+      override def taskSucceeded(partition: Int, value: Any): Unit = numResults += 1
+      override def jobFailed(exception: Exception): Unit = {
+        failureReason = Some(exception)
+      }
     }
     val jobId = submit(new MyRDD(sc, 0, Nil), Array(), listener = fakeListener)
     assert(numResults === 0)
     cancel(jobId)
+    assert(failureReason.isDefined)
+    assert(failureReason.get.getMessage() === "Job 0 cancelled ")
```
Comment (squito) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: this test used to just log an exception on `cancel(jobId)`.  I'm not sure what it was supposed to be testing before.  I made the minimal change here, by capturing the exception and checking it.  But maybe `cancel(jobId)` should *not* be creating an exception?  Is the idea that if you submit a job with no partitions, it will immediately stop?  That way, if you try to cancel it, you'd just hit [this case](https://github.com/apache/spark/blob/bb1640529725c6c38103b95af004f8bd90eeee5c/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala#L1224) with a harmless `logDebug`?  That suggests we should change [`handleJobSubmitted`](https://github.com/apache/spark/blob/bb1640529725c6c38103b95af004f8bd90eeee5c/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala#L757) to handle empty jobs, the same way we handle it in [`submitMissingTasks`](https://github.com/apache/spark/blob/bb1640529725c6c38103b95af004f8bd90eeee5c/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala#L910) for stages with no partitions.

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 4]
```diff
@@ -700,7 +700,8 @@ class DAGScheduler(
 
   private[scheduler] def cleanUpAfterSchedulerStop() {
     for (job <- activeJobs) {
-      val error = new SparkException("Job cancelled because SparkContext was shut down")
+      val error =
+        new SparkException(s"Job ${job.jobId} cancelled because SparkContext was shut down")
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: not necessary, but was helpful for some debugging, figured it couldn't hurt.

**Comment by mateiz** [comment]
Hey, so is the conclusion that the DAGScheduler actually did pass the exception to JobListeners, but we weren't listening for it in our test suite? I thought the initial problem was that some exception got swallowed before ever being passed up.

**Comment by squito** [comment]
Hi @mateiz sorry I totally missed your comment a while back. so I think poorly explained the issue in the first place.  Exceptions are already making it to the joblisteners to the best of my knowledge (as part of [`doCancelAllJobs`](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala#L1648)).

The issue that I'm trying to address is just confusing behavior in the tests.  Before this change, if there is an exception inside the event processing loop, then all that happens is [some internal state is set](https://github.com/apache/spark/blob/master/core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala#L172).  Furthermore, the event process loop [stops the `SparkContext`](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala#L1652), but it doesn't actually stop the `DAGScheduler` in use b/c in these tests the [`DAGScheduler` is custom, not the one created by the `SparkContext`](https://github.com/apache/spark/blob/master/core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala#L195).

The net effect is that when a test doesn't behave the way you expect it to, its hard to unravel why.  Unless tests include `assert(failure == null)` on every other line, if you do hit an exception, the test will keep running quite a bit further but with strange behavior which can be hard to make sense of.

The only goal here is just to make those issues much more immediately obvious.  For these tests, as soon as there is an exception in the event process loop, throw the exception and fail the test.

**Comment by vanzin** [comment]
LGTM.

**Comment by zsxwing** [comment]
retest this please

**Comment by zsxwing** [comment]
LGTM pending tests

**Comment by andrewor14** [comment]
@squito looks like this is failing a legit test

**Comment by andrewor14** [comment]
retest this please

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 5]
```diff
@@ -1290,7 +1291,7 @@ class DAGScheduler(
       case TaskResultLost =>
         // Do nothing here; the TaskScheduler handles these failures and resubmits the task.
 
-      case other =>
+      case _: ExecutorLostFailure | TaskKilled | UnknownReason =>
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: is this a related change?

Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: If so, I think the compiler will report uncompleted matching as a warning and the build will fail. Right?

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: I see, that's fine

Comment (squito) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: this was needed just for the added test case -- I needed a way to intentionally throw an error inside the event loop, and with this change if `event.reason` is `null`, you get an exception.

And yeah, if you add another type but dont' include it in this match, you'll get a fatal warning in compilation since its a sealed trait.

**Comment by squito** [comment]
I don't think that test failure is related at all, but I also dont' see why that test would be flaky :(

**Comment by squito** [comment]
Jenkins, retest this please

**Comment by andrewor14** [comment]
retest this please

**Comment by andrewor14** [comment]
Test failure is unrelated. Merging into master.
